File: hq-memes/frontend/src/app/core/services/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

import { environment } from '../../../environments/environment';
import { ProfileUpdateRequest, UserSummary } from '../models/user.model';
import { MessageResponse } from '../models/api-response.model';

@Injectable({
  providedIn: 'root'
})
export class UserService {
  private apiUrl = `${environment.apiUrl}/users`;

  constructor(private http: HttpClient) { }

  getCurrentUser(): Observable<UserSummary> {
    return this.http.get<UserSummary>(`${this.apiUrl}/me`);
  }

  updateProfile(profileUpdateRequest: ProfileUpdateRequest): Observable<MessageResponse> {
    return this.http.put<MessageResponse>(`${this.apiUrl}/me`, profileUpdateRequest);
  }

  updateProfilePicture(file: File): Observable<MessageResponse> {
    const formData = new FormData();
    formData.append('file', file);
    return this.http.post<MessageResponse>(`${this.apiUrl}/me/profile-picture`, formData);
  }
}

File: backend/src/main/java/com/memevote/backend/service/UserService.java
package com.memevote.backend.service;

import com.memevote.backend.dto.request.ProfileUpdateRequest;
import com.memevote.backend.dto.response.MessageResponse;
import com.memevote.backend.dto.response.UserSummary;
import com.memevote.backend.model.User;
import com.memevote.backend.repository.UserRepository;
import com.memevote.backend.security.services.UserDetailsImpl;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;

@Service
public class UserService {
    @Autowired
    private UserRepository userRepository;

    @Autowired
    private ImageStorageService imageStorageService;

    public User getCurrentUser() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        if (authentication == null || !(authentication.getPrincipal() instanceof UserDetailsImpl)) {
            throw new RuntimeException("User not authenticated");
        }

        UserDetailsImpl userDetails = (UserDetailsImpl) authentication.getPrincipal();
        return userRepository.findById(userDetails.getId())
                .orElseThrow(() -> new RuntimeException("User not found"));
    }

    public UserSummary getUserSummary(User user) {
        return new UserSummary(user.getId(), user.getUsername(), user.getEmail(), user.getProfilePicture());
    }

    public MessageResponse updateProfile(ProfileUpdateRequest profileUpdateRequest) {
        User currentUser = getCurrentUser();

        if (profileUpdateRequest.getUsername() != null &&
            !profileUpdateRequest.getUsername().equals(currentUser.getUsername())) {

            if (userRepository.existsByUsername(profileUpdateRequest.getUsername())) {
                return new MessageResponse("Error: Username is already taken!");
            }

            currentUser.setUsername(profileUpdateRequest.getUsername());
            userRepository.save(currentUser);
        }

        return new MessageResponse("Profile updated successfully!");
    }

    public MessageResponse updateProfilePicture(MultipartFile file) throws IOException {
        User currentUser = getCurrentUser();

        // Delete old profile picture if exists
        if (currentUser.getProfilePicture() != null) {
            imageStorageService.deleteImage(currentUser.getProfilePicture());
        }

        // Store new profile picture
        String fileName = imageStorageService.storeImage(file);
        currentUser.setProfilePicture(fileName);
        userRepository.save(currentUser);

        return new MessageResponse("Profile picture updated successfully!");
    }
}
